virtual_ecosystem/models/animal/plant_resources_v3.py

```diff
@@ -0,0 +1,344 @@
+        cell_id=cell_ids,
+        pft=pfts,
+    ),
+)
+
+elemental_ratio = DataArray(
+    np.ones((data.grid.n_cells, pfts.size)),
+    dims=("cell_id", "pft"),
+    coords=dict(
+        cell_id=cell_ids,
+        pft=pfts,
+    ),
+)
+
+data.add_from_dict(
+    dict(
+        leaf_mass=mass,
+        leaf_n_ratio=elemental_ratio + 0.05,
+        leaf_p_ratio=elemental_ratio + 0.02,
+        leaf_mass_consumed=elemental_ratio,
+        subcanopy_mass=mass[0, :, 0] * 3,
+        subcanopy_n_ratio=elemental_ratio[:, 0] + 0.03,
+        subcanopy_p_ratio=elemental_ratio[:, 0] + 0.01,
+        subcanopy_mass_consumed=elemental_ratio[:, 0],
+    )
+)
+
+# Get a tuple of all herbivory array resources
+
+leaves = [
+    ArrayResources(
+        data=data,
+        mass_var="leaf_mass",
+        n_ratio_var="leaf_n_ratio",
+        p_ratio_var="leaf_p_ratio",
+        mass_consumed_var="leaf_mass_consumed",
+        vertical_occupancy=VerticalOccupancy.CANOPY,
+        pft=pft,
+        vertical_layers=DataArray([0, 1, 2], dims="vertical_layer"),
+    )
+    for pft in data["pft"].to_numpy()
+]
+
+subcanopy = ArrayResources(
+    data=data,
+    mass_var="subcanopy_mass",
+    n_ratio_var="subcanopy_n_ratio",
+    p_ratio_var="subcanopy_p_ratio",
+    mass_consumed_var="subcanopy_mass_consumed",
+    vertical_occupancy=VerticalOccupancy.GROUND,
+)
+
+
+# Faked consumer namespace and hence a list of four consumers
+consumer = SimpleNamespace(
+    functional_group=SimpleNamespace(
+        mechanical_efficiency=0.9, conversion_efficiency=0.6
+    )
+)
+
+consumers = [consumer] * 4
+
+rng = np.random.default_rng()
+
+# At init, create a list of herbivory resources:
+
+herbivory_resources = (*leaves, subcanopy)
+print(herbivory_resources)
+
+# Then within a single time loop
+for cell_id in np.arange(12):
+    # Get the cell level resource for this cell from each array resource
+    cell_resources = [array_res[cell_id] for array_res in herbivory_resources]
+
+    # All consumers feed on all four resources
+    for herbivore in [consumer] * 4:
+        for resource in cell_resources:
+            resource.get_eaten(consumed_mass=rng.random(), consumer=herbivore)
+
+# After the loop finishes print out the accumulated herbivory in the ArrayResource
+# instances
+for array_res in herbivory_resources:
+    print(array_res)
+    print(array_res.consumed_mass)
+
+# The data object still doesn't know about herbivory
+data["leaf_mass_consumed"]
+data["subcanopy_mass_consumed"]
+
+# So use the write_herbivory() to push the accumualate value out to the Data object.
+# Note that the PFT specific resources are writing to different parts of the same data
+# array.
+for array_res in herbivory_resources:
+    array_res.write_herbivory()
+
+data["leaf_mass_consumed"]
+data["subcanopy_mass_consumed"]
+
+# And then, before starting the next loop:
+
+for array_res in herbivory_resources:
+    array_res.set_mass_and_elemental_ratios()
```

Comment:
Where would all of these be happening? Within the `AnimalModel` `update` or somewhere else?

Also, that triply nested `for` loop looks super inefficient. although it is not obvious how to improve it as none of the functionality is vectorised...

Reply:
From this line down, it would be within `AnimalModel._update` (although probably as  a call to an encapsulating method?):

> # Get a tuple of all herbivory array resources

Everything above that is the PlantsModel pushing Plant data into the `Data` object,  in `PlantsModel.__init__` and `PlantsModel._update`